### DISCUSSION TITLE
could add some VIM-like keymaps in ulauncher?

### DISCUSSION THREAD

**Discussion by sunn4room** [Ulauncher/Ulauncher#909] [Ideas]
vim-like keymap affected me a lot, include vscode, idea...  I think ulauncher could be better if it also have a vim-like keymap. example:
- type jj when input, can change focus on items
- type j and k, can move focus up and down
- type gg, go first
- type G, go last

please...

**Comment by troycurtisjr** [comment]
Currently hitting ESC would close ulauncher, in your proposed functionality, would that put you into "normal mode"? Or are you thinking of just using the "quick type of `jj`" to change focus. And if so, then should that be customizable? After all, I use `kj` instead, and while `jj` is common, it isn't universal (unlike ESC for instance). If there *was* a normal mode, would we have to implement at least a basic set of editing commands to make it feel natural? 

This seems like it could be a good deal of code in order to support. Even as a heavy vim user, and someone that has vimium installed in every browser I use, I'm not sure it really works well in this context. My alternative, that I plan to post a PR for soonish, is to implement the standard command-line shortcuts (and emacs) C-n and C-p for moving up and down the list. (#649 ). I used the vi mode in my shell for the longest time, but it doesn't work *quite* as well in that context as for editing a bunch of text files. I found that learning the basic emacs style bindings for use on the shell was a lot more productive. Obviously just a personal preference!

**Comment by sunn4room** [comment]
#649 is a good idea. And each has advantages and disadvantages.
vim style bindings could avoid to hit `control` or `meta`, but there is a little delay to wait next `j` after type first `j`.
emacs style bindings is perfect for emacs user, but as a vim user, I don't like to hite `control` for moving up and down.
After all, both proposals are acceptable, and is better than use arrow key for moving up and down

**Comment by Nimmidev** [comment]
I think it would already be a huge improvement if you could navigate search results with Ctrl-j and Ctrl-k

**Reply by troycurtisjr** [reply]
Note that you can use C-n (down/next) & C-p (up/previous). Of course those are emacs bindings, not vim 😉. We talked about various bindings, but there are potentially many conflicting ones so they'd need to be all optional and configurable. That means it's a pretty big effort to implement and to maintain.

**Reply by nelsonmestevao** [reply]
Is it normal that C-a (go to beginning) and C-e (go to end) don't work the same as emacs?

**Reply by waltermas-gitter** [reply]
@troycurtisjr Could I change C-n, C-p to C-j, C-k on my config? How can I do that? Thanks!

**Reply by troycurtisjr** [reply]
> Is it normal that C-a (go to beginning) and C-e (go to end) don't work the same as emacs?

ulauncher doesn't do any special input form edit handling, it is simply a GTK widget. If you want emacs style line-editing, you will need to enable the emacs keymap in Gnome settings. The up/down is a special case since it is controlling functionality outside of the input widget, and thus must be implemented by ulauncher.

**Reply by sunn4room** [reply]
If someone want more custom key bindings and doesn't exclude TUI program, it's a good choice to use `fzf` in terminal emulator. `fzf` is a great search tool, and it supports custom key bindings.

[It's my attempt](https://github.com/sunn4room/dotfiles/blob/main/dot_local/bin/executable_fzflch)
